zokrates_core/src/static_analysis/constant_inliner.rs

```diff
@@ -0,0 +1,856 @@
+use crate::typed_absy::folder::*;
+use crate::typed_absy::*;
+use std::convert::TryInto;
+use zokrates_field::Field;
+
+pub struct ConstantInliner<'ast, T: Field> {
+    modules: TypedModules<'ast, T>,
+    location: OwnedTypedModuleId,
+}
+
+impl<'ast, T: Field> ConstantInliner<'ast, T> {
+    fn with_modules_and_location(
+        modules: TypedModules<'ast, T>,
+        location: OwnedTypedModuleId,
+    ) -> Self {
+        ConstantInliner { modules, location }
+    }
+
+    pub fn inline(p: TypedProgram<'ast, T>) -> TypedProgram<'ast, T> {
+        // initialize an inliner over all modules, starting from the main module
+        let mut inliner =
+            ConstantInliner::with_modules_and_location(p.modules.clone(), p.main.clone());
+
+        inliner.fold_program(p)
+    }
+
+    pub fn module(&self) -> &TypedModule<'ast, T> {
+        self.modules.get(&self.location).unwrap()
+    }
+
+    pub fn change_location(&mut self, location: OwnedTypedModuleId) -> OwnedTypedModuleId {
+        let prev = self.location.clone();
+        self.location = location;
+        prev
+    }
+
+    pub fn get_constant(&mut self, id: &Identifier) -> Option<TypedConstant<'ast, T>> {
+        self.modules
+            .get(&self.location)
+            .unwrap()
+            .constants
+            .as_ref()
+            .and_then(|c| c.get(id.clone().try_into().unwrap()))
+            .cloned()
+            .and_then(|tc| {
+                let symbol = self.fold_constant_symbol(tc);
+                match symbol {
+                    TypedConstantSymbol::Here(tc) => Some(tc),
+                    _ => None,
```

Comment:
Same here, is this case reachable?

Reply:
`fold_constant_symbol` recursively finds `Here` so in the example where we import an imported constant, it does the recursion until we are in the module that actually defined the constant (`There` -> `There` -> `Here`). So in this case, I the default arm should be `unreachable`, not `None`.

Reply:
I see. In order to avoid unreachable match arms, maybe it would make sense to extract the canonicalisation process to return what's inside `Here`, rather than the `TypedConstantSymbol`. Then the runtime assumption is that any import chains ends in a `Here`, which to me sounds clearer than the runtime assumption that the `There` match arm is unreachable.

Reply:
This was not possible in the function call inliner, because the resolving process can return either `Here` or `Flat`, and representing these two options would require defining another enum, which felt like overkill, but in the case of constants there's no Flat so I feel it would work?